tests/v2/validation/schemas/schemas.go

```diff
@@ -0,0 +1,294 @@
+}
+
+func getAPIGroupInfoByAPIGroupName(client *rancher.Client, clusterID, apiGroupName string) (map[string]interface{}, error) {
+	return getJSONResponse(client, clusterID, apiGroupEndpoint, apiGroupName)
+}
+
+func accessSchemaDefinitionForEachSchema(client *rancher.Client, schemasCollection *v1.SteveCollection, clusterID string) ([]string, error) {
+	var failedSchemaDefinitions []string
+	var err error
+	var schemaID string
+
+	for _, schema := range schemasCollection.Data {
+		schemaID = schema.JSONResp["id"].(string)
+
+		if _, exists := exceptionMap[schemaID]; !exists {
+			_, err = getSchemaByID(client, clusterID, schemaID)
+			if err != nil {
+				return nil, err
+			}
+			_, err = getSchemaDefinitionByID(client, clusterID, schemaID)
+			if err != nil {
+				failedSchemaDefinitions = append(failedSchemaDefinitions, schemaID)
+			}
+		}
+	}
+	return failedSchemaDefinitions, nil
+}
+
+func checkPreferredVersion(client *rancher.Client, schemasCollection *v1.SteveCollection, clusterID string) ([][]string, error) {
+	var failedSchemaPreferredVersionCheck [][]string
+
+	for _, schema := range schemasCollection.Data {
+		schemaID := schema.JSONResp["id"].(string)
+
+		if _, exists := exceptionMap[schemaID]; !exists {
+			schemaInfo, err := getSchemaByID(client, clusterID, schemaID)
+			if err != nil {
+				return nil, err
+			}
+			schemaVersion := schemaInfo["attributes"].(map[string]interface{})["version"].(string)
+
+			apiGroupName := schemaInfo["attributes"].(map[string]interface{})["group"].(string)
+			if apiGroupName == "" {
+				continue
+			}
+			apiGroupInfo, err := getAPIGroupInfoByAPIGroupName(client, clusterID, apiGroupName)
+			if err != nil {
+				return nil, err
+			}
+			preferredVersion := apiGroupInfo["preferredVersion"].(map[string]interface{})["version"].(string)
+
+			if schemaVersion != preferredVersion {
+				failedSchemaPreferredVersionCheck = append(failedSchemaPreferredVersionCheck, []string{
+					"Schema ID: " + schemaID + ",",
+					"Schema Version: " + schemaVersion + ";",
+					"API Group Name: " + apiGroupName + ",",
+					"API Group Version: " + preferredVersion,
+				})
+			}
+		}
+	}
+	return failedSchemaPreferredVersionCheck, nil
+}
+
+func getJSONResponse(client *rancher.Client, clusterID, endpointType, existingID string) (map[string]interface{}, error) {
+	rancherURL := client.RancherConfig.Host
+	token := client.RancherConfig.AdminToken
+
+	var httpURL string
+	if clusterID == "local" {
+		if endpointType == apiGroupEndpoint {
+			httpURL = fmt.Sprintf("https://%s/apis/%s", rancherURL, existingID)
+		} else {
+			httpURL = fmt.Sprintf("https://%s/v1/%s/%s", rancherURL, endpointType, existingID)
+		}
+	} else {
+		if endpointType == apiGroupEndpoint {
+			httpURL = fmt.Sprintf("https://%s/k8s/clusters/%s/apis/%s", rancherURL, clusterID, existingID)
+		} else {
+			httpURL = fmt.Sprintf("https://%s/k8s/clusters/%s/v1/%s/%s", rancherURL, clusterID, endpointType, existingID)
+		}
+	}
```

Comment:
```suggestion
baseURL := fmt.Sprintf("https://%s", rancherURL)

if clusterID != "local" {
    baseURL = fmt.Sprintf("%s/k8s/clusters/%s", baseURL, clusterID)
}

if endpointType == apiGroupEndpoint {
    httpURL = fmt.Sprintf("%s/apis/%s", baseURL, existingID)
} else {
    httpURL = fmt.Sprintf("%s/v1/%s/%s", baseURL, endpointType, existingID)
}
```

^^ to reduce the conditional statements.

Reply:
Great point :)